client/network/src/service.rs

```diff
@@ -716,34 +775,40 @@ impl<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> Stream for Ne
 					self.network_service.add_known_address(peer_id, addr),
 				ServerToWorkerMsg::SyncFork(peer_ids, hash, number) =>
 					self.network_service.user_protocol_mut().set_sync_fork_request(peer_ids, &hash, number),
+				ServerToWorkerMsg::EventsStream(sender) =>
+					self.event_streams.push(sender),
+				ServerToWorkerMsg::WriteNotification { message, proto_name, target } =>
+					self.network_service.user_protocol_mut().write_notification(target, proto_name, message),
+				ServerToWorkerMsg::RegisterNotifProtocol { proto_name, engine_id, handshake_msg } =>
+					self.network_service.user_protocol_mut()
+						.register_notif_protocol(proto_name, engine_id, handshake_msg),
+				ServerToWorkerMsg::DisconnectPeer(who) =>
+					self.network_service.user_protocol_mut().disconnect_peer(&who),
 			}
 		}
 
 		loop {
 			// Process the next action coming from the network.
 			let poll_value = self.network_service.poll();
 
-			let outcome = match poll_value {
+			match poll_value {
 				Ok(Async::NotReady) => break,
-				Ok(Async::Ready(Some(BehaviourOut::SubstrateAction(outcome)))) => outcome,
-				Ok(Async::Ready(Some(BehaviourOut::Dht(ev)))) =>
-					return Ok(Async::Ready(Some(Event::Dht(ev)))),
-				Ok(Async::Ready(None)) => CustomMessageOutcome::None,
+				Ok(Async::Ready(Some(BehaviourOut::BlockImport(origin, blocks)))) =>
+					self.import_queue.import_blocks(origin, blocks),
+				Ok(Async::Ready(Some(BehaviourOut::JustificationImport(origin, hash, nb, justification)))) =>
+					self.import_queue.import_justification(origin, hash, nb, justification),
+				Ok(Async::Ready(Some(BehaviourOut::FinalityProofImport(origin, hash, nb, proof)))) =>
+					self.import_queue.import_finality_proof(origin, hash, nb, proof),
+				Ok(Async::Ready(Some(BehaviourOut::Event(ev)))) => {
+					self.event_streams.retain(|sender| sender.unbounded_send(ev.clone()).is_ok());
```

Comment:
Follow up to https://github.com/paritytech/substrate/pull/4125#discussion_r348355881 (can't comment there no more).

Either a `Bytes` or we make the protocol name available here so we can filter early on?